src/mlpack/core/optimizers/adagrad/ada_grad.hpp

```diff
@@ -0,0 +1,155 @@
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef __MLPACK_CORE_OPTIMIZERS_ADAGRAD_ADA_GRAD_HPP
+#define __MLPACK_CORE_OPTIMIZERS_ADAGRAD_ADA_GRAD_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace optimization {
+
+/**
+ * Adagrad is a modified version of stochastic gradient descent which performs
+ * larger updates for more sparse parameters and smaller updates for less sparse
+ * parameters.
+ *
+ * For more information, see the following.
+ *
+ * @code
+ * @article{duchi2011adaptive,
+ *   author    = {Duchi, John and Hazan, Elad and Singer, Yoram},
+ *   title     = {Adaptive subgradient methods for online learning and stochastic optimization},
+ *   journal   = {Journal of Machine Learning Research},
+ *   year      = {2011}
+ * }
+ * @endcode
+ *
+
+ * For AdaGrad to work, a DecomposableFunctionType template parameter is
+ * required. This class must implement the following function:
+ *
+ *   size_t NumFunctions();
+ *   double Evaluate(const arma::mat& coordinates, const size_t i);
+ *   void Gradient(const arma::mat& coordinates,
+ *                 const size_t i,
+ *                 arma::mat& gradient);
+ *
+ * NumFunctions() should return the number of functions (\f$n\f$), and in the
+ * other two functions, the parameter i refers to which individual function (or
+ * gradient) is being evaluated.  So, for the case of a data-dependent function,
+ * such as NCA (see mlpack::nca::NCA), NumFunctions() should return the number
+ * of points in the dataset, and Evaluate(coordinates, 0) will evaluate the
+ * objective function on the first point in the dataset (presumably, the dataset
+ * is held internally in the DecomposableFunctionType).
+ *
+ * @tparam DecomposableFunctionType Decomposable objective function type to be
+ *     minimized.
+ */
+template<typename DecomposableFunctionType>
+class AdaGrad
+{
+ public:
+  /**
+   * Construct the AdaGrad optimizer with the given function and parameters.
+   * The defaults here are not necessarily good for the given problem, so it is
+   * suggested that the values used be tailored to the task at hand. The
+   * maximum number of iterations refers to the maximum number of points that
+   * are processed (i.e., one iteration equals one point; one iteration does not
+   * equal one pass over the dataset).
+   *
+   * @param function Function to be optimized (minimized).
+   * @param stepSize Step size for each iteration
+   * @param eps Value used to initialise the squared gradient parameter.
+   * @param maxIterations Maximum number of iterations allowed (0 means no
+   *        limit).
+   * @param tolerance Maximum absolute tolerance to terminate algorithm.
+   * @param shuffle If true, the function order is shuffled; otherwise, each
+   *        function is visited in linear order.
+   */
+  AdaGrad(DecomposableFunctionType& function,
+      const double stepSize = 0.01,
+      const double eps = 1e-8,
```

Comment:
The name `eps` may confuse users, who might think that `eps` is the tolerance before termination.  I might consider a different name here.